### PR TITLE
[ContactStructuralMechanicsApplicatiuon] Adding MixedULMSolver

### PR DESCRIPTION
**Pull request by loumalouomega**
This PR add the MixedULMSolver. The intention of this PR is to focus just in this feature, and related utilities

A python interface and a CPP test is added.

Sparse matrix utility with corresponding test is added

After this is merged a full PR contaning the new element necessary to use it and the corresponding tests will prepared

### CODE REVIEW

File: applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
-from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+from __future__ import print_function, absolute_import, division  # makes KM backward compatible with python 2.6 and 2.7
 #import kratos core and applications
-import KratosMultiphysics
-import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
-import KratosMultiphysics.ContactStructuralMechanicsApplication as ContactStructuralMechanicsApplication
+import KratosMultiphysics as KM
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py:3]: I think I like your new syntax
Avoids the long names and also the opportunity to use different abbreviations (e.g.
- `import KratosMultiphysics.StructuralMechanicsApplication`
- `import KratosMultiphysics.StructuralMechanicsApplication as KratosStructural`)
- ...

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 2]
```diff
@@ -0,0 +1,2032 @@
...
+
+// External includes
+#include <boost/numeric/ublas/vector.hpp>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "linear_solvers/reorderer.h"
+#include "linear_solvers/iterative_solver.h"
+#include "solving_strategies/builder_and_solvers/builder_and_solver.h"
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: you shouldn't need to include the builder and solver

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 3]
```diff
@@ -0,0 +1,2032 @@
...
+        // Get the u and lm residuals
+        GetUPart (rB, mResidualDisp);
+        
+        KRATOS_DETAIL("mKDispModified") << mKDispModified << std::endl;
+        KRATOS_DETAIL("mResidualDisp") << mResidualDisp << std::endl;
+        
+        // Solve u block
+        if (mDisp.size() != total_size)
+            mDisp.resize(total_size, false);
+        mpSolverDispBlock->Solve (mKDispModified, mDisp, mResidualDisp);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: mmm here mDisp will have a different size than mKDispModified no?

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 4]
```diff
@@ -0,0 +1,2032 @@
...
+     * @brief This function is designed to clean up all internal data in the solver.
+     * @details Clear is designed to leave the solver object as if newly created. After a clear a new Initialize is needed
+     */
+    void Clear() override
+    {        
+        mBlocksAreAllocated = false;
+        mpSolverDispBlock->Clear();
+        
+        // We clear the matrixes and vectors
+        mKDispModified.clear(); /// The modified displacement block
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: take care, this only sets it to zero, does not deallocate

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 5]
```diff
@@ -0,0 +1,2032 @@
...
+     * @param rX Solution vector. it's also the initial guess for iterative linear solvers.
+     * @param rB Right hand side vector.
+     */
+    bool Solve(
+        SparseMatrixType& rA, 
+        VectorType& rX, 
+        VectorType& rB
+        ) override
+    {
+        if (mIsInitialized == false)
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: what happens if you change the system's shape?

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 6]
```diff
@@ -0,0 +1,2032 @@
...
+        SizeType n_lm_inactive_dofs = 0, n_lm_active_dofs = 0;
+        SizeType n_master_dofs = 0;
+        SizeType n_slave_inactive_dofs = 0, n_slave_active_dofs = 0;
+        SizeType tot_active_dofs = 0;
+        for (auto& i_dof : rDofSet) {
+            node_id = i_dof.Id();
+            pnode = rModelPart.pGetNode(node_id);
+            if (i_dof.EquationId() < rA.size1()) {
+                tot_active_dofs++;
+                if (i_dof.GetVariable().Key() == VECTOR_LAGRANGE_MULTIPLIER_X ||
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: why don't you compare the variable directly?

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 7]
```diff
@@ -0,0 +1,2032 @@
...
+            if (i_dof.EquationId() < rA.size1()) {
+                tot_active_dofs++;
+                if (i_dof.GetVariable().Key() == VECTOR_LAGRANGE_MULTIPLIER_X || 
+                    i_dof.GetVariable().Key() == VECTOR_LAGRANGE_MULTIPLIER_Y || 
+                    i_dof.GetVariable().Key() == VECTOR_LAGRANGE_MULTIPLIER_Z) {
+                    if (pnode->Is(ACTIVE))
+                        n_lm_active_dofs++;
+                    else
+                        n_lm_inactive_dofs++;
+                } else if (i_dof.GetVariable().Key() == DISPLACEMENT_X ||
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: i would say that the if is unnecessary

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 8]
```diff
@@ -0,0 +1,2032 @@
...
+                        }
+                    }
+                }
+            }
+        }
+
+        KRATOS_ERROR_IF(tot_active_dofs != rA.size1()) << "Total system size does not coincide with the free dof map" << std::endl;
+
+        // Resize arrays as needed
+        mMasterIndices.resize (n_master_dofs,false);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: check the sizes before resizing. They may be already ok

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 9]
```diff
@@ -0,0 +1,2032 @@
...
+         * mGlobalToLocalIndexing[i] will contain the position in the local blocks of the
+         */
+        SizeType lm_inactive_counter = 0, lm_active_counter = 0;
+        SizeType master_counter = 0;
+        SizeType slave_inactive_counter = 0, slave_active_counter = 0;
+        SizeType other_counter = 0;
+        IndexType global_pos = 0;
+        for (auto& i_dof : rDofSet) {
+            node_id = i_dof.Id();
+            pnode = rModelPart.pGetNode(node_id);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:513]: this has a logarithmic cost. not sure if you could merge this loop with the one before...

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h [hunk 10]
```diff
@@ -0,0 +1,2032 @@
...
+        IndexType* mKSASA_ptr = new IndexType[slave_active_size + 1];
+        IndexType* KSALMA_ptr = new IndexType[slave_active_size + 1];
+        IndexType* KLMILMI_ptr = new IndexType[lm_inactive_size + 1];
+        IndexType* KLMALMA_ptr = new IndexType[lm_active_size + 1];
+        
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(master_size + 1); i++) 
+            KMLMA_ptr[i] = 0;
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(slave_active_size + 1); i++) {
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: do you need setting this to zero?

**Review by RiccardoRossi — DISMISSED** [review]

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 2 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: Ok

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: No, total_size is total displacement size, maybe I should change the name

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 5 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: I am not having problems for that, and I change constantly the size of the system

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 4 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: OK

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 6 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: I took the code from another place...

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 7 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: Nope, imagine that I have master and slave with thermal DoF

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 8 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: Ok

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 9 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:513]: Maybe I can build a map and count from there

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h (hunk 10 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h]: Yes, some matrices have empty rows... I assure you this was the only way to make it work always in full debug

File: applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py [hunk 11]
```diff
@@ -173,6 +173,26 @@ def _create_linear_solver(self):
...
+                            "smoother_type":"ilu0",
+                            "krylov_type": "lgmres",
+                            "coarsening_type": "aggregation",
+                            "max_iteration": 100,
+                            "provide_coordinates": false,
+                            "gmres_krylov_space_dimension": 100,
+                            "verbosity" : 1,
+                            "tolerance": 1e-6,
+                            "scaling": false,
+                            "block_size": 3,
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py:190]: i am approving, however please note that if you are in 2D the block size should be 2

File: applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_static_solver.py [hunk 12]
```diff
@@ -198,18 +198,38 @@ def _create_linear_solver(self):
                 KM.Logger.PrintWarning("Using MixedULMLinearSolver, definition of ALM parameters recommended")
                 name_mixed_solver = self.contact_settings["mixed_ulm_solver_parameters"]["solver_type"].GetString()
                 if (name_mixed_solver == "MixedULMLinearSolver"):
+                    linear_solver_name = self.settings["linear_solver_settings"]["solver_type"].GetString()
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_static_solver.py]: why this? it should detect this internally

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_static_solver.py]: The current linear solver factory lacks that capability, we should push #615

File: applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py (hunk 11 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py:190]: True

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py:190]: Done

File: applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py [hunk 13]
```diff
@@ -45,7 +45,14 @@ def __init__(self, main_model_part, custom_settings):
                 "contact_displacement_relative_tolerance": 1.0e-4,
                 "contact_displacement_absolute_tolerance": 1.0e-9,
                 "contact_residual_relative_tolerance"    : 1.0e-4,
-                "contact_residual_absolute_tolerance"    : 1.0e-9
+                "contact_residual_absolute_tolerance"    : 1.0e-9,
+                "use_mixed_ulm_solver"                   : true,
+                "mixed_ulm_solver_parameters" :
+                {
+                    "solver_type": "MixedULMLinearSolver",
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py]: I recommend using a style-guide conforming name: `mixed_ulm_linear_solver`

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_implicit_dynamic_solver.py]: Done

**Comment by loumalouomega** [comment]
Can your reapprove @RiccardoRossi ?

**Review by RiccardoRossi — APPROVED** [review]

**Comment by loumalouomega** [comment]
Thanks!!

**Pull request merged**